terracotta/drivers/geotiff_raster_store.py

```diff
@@ -0,0 +1,164 @@
+from concurrent.futures.process import BrokenProcessPool
+
+import functools
+import logging
+import warnings
+import threading
+
+import numpy as np
+
+from terracotta import get_settings
+from terracotta import raster
+from terracotta.cache import CompressedLFUCache
+from terracotta.drivers.base_classes import RasterStore
+
+Number = TypeVar('Number', int, float)
+
+logger = logging.getLogger(__name__)
+
+context = threading.local()
+context.executor = None
+
+
+def create_executor() -> Executor:
+    settings = get_settings()
+
+    if not settings.USE_MULTIPROCESSING:
+        return ThreadPoolExecutor(max_workers=1)
+
+    executor: Executor
+
+    try:
+        # this fails on architectures without /dev/shm
+        executor = ProcessPoolExecutor(max_workers=3)
+    except OSError:
+        # fall back to serial evaluation
+        warnings.warn(
+            'Multiprocessing is not available on this system. '
+            'Falling back to serial execution.'
+        )
+        executor = ThreadPoolExecutor(max_workers=1)
+
+    return executor
+
+
+def submit_to_executor(task: Callable[..., Any]) -> Future:
+    if context.executor is None:
+        context.executor = create_executor()
+
+    try:
+        future = context.executor.submit(task)
+    except BrokenProcessPool:
+        # re-create executor and try again
+        logger.warn('Re-creating broken process pool')
+        context.executor = create_executor()
+        future = context.executor.submit(task)
+
+    return future
+
+
+class GeoTiffRasterStore(RasterStore):
+    """Mixin that implements methods to load raster data from disk.
+
+    get_datasets has to return path to raster file as sole dict value.
+    """
+    _TARGET_CRS: str = 'epsg:3857'
+    _LARGE_RASTER_THRESHOLD: int = 10980 * 10980
+    _RIO_ENV_OPTIONS = dict(
+        GDAL_TIFF_INTERNAL_MASK=True,
+        GDAL_DISABLE_READDIR_ON_OPEN='EMPTY_DIR'
+    )
+
+    def __init__(self) -> None:
+        settings = get_settings()
+        self._raster_cache = CompressedLFUCache(
+            settings.RASTER_CACHE_SIZE,
+            compression_level=settings.RASTER_CACHE_COMPRESS_LEVEL
+        )
+        self._cache_lock = threading.RLock()
+
+        from rasterio import Env
+        self._rio_env = Env(**self._RIO_ENV_OPTIONS)
+
+    def compute_metadata(self, path: str, *,
+                         extra_metadata: Any = None,
+                         use_chunks: bool = None,
+                         max_shape: Sequence[int] = None) -> Dict[str, Any]:
+        return raster.compute_metadata(path, extra_metadata=extra_metadata,
+                                       use_chunks=use_chunks, max_shape=max_shape,
+                                       large_raster_threshold=self._LARGE_RASTER_THRESHOLD,
+                                       rio_env=self._rio_env)
+
+    # return type has to be Any until mypy supports conditional return types
```

Comment:
I think this is just
```python
from typing import Union

def compute_metadata(...) -> Union[Future[np.ma.MaskedArray], np.ma.MaskedArray]
```

But in any case, having multiple return types sort of defeats the whole purpose of static types.
I can't see how a statically typed language could ever allow something like this without generics or metaprogramming.

The proper thing to do here would probably be to split this into 2 methods `compute_metadata` and `compute_metadata_async`. But I guess this might break the API elsewhere.

This is actual a general problem with async, that you basically have to implement your API twice if you want to have both an async and non-async API. There is a semi-famous rant about this problem here http://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/

Reply:
Anyways, if you are strict about mypy/static type hints, then using `Union` is usually a bit of a code smell.

Only exception is `Optional[SomeType]` of course, which is just shorthand for `Union[SomeType, None]`.

Reply:
`Union` means "type could be either this or that", so you have to guard against both cases in downstream code. However, in this case, the return type is perfectly deterministic, based on the `async` input value.

> I can't see how a statically typed language could ever allow something like this without generics or metaprogramming.

```python
@typing.overload
def compute_metadata(..., async: TrueType) -> Future[np.ma.MaskedArray]:
    pass

@typing.overload
def compute_metadata(..., async: FalseType) -> np.ma.MaskedArray:
    pass
```

(pretty clean IMO, but not supported)

> The proper thing to do here would probably be to split this into 2 methods compute_metadata and compute_metadata_async. But I guess this might break the API elsewhere.

The code is clear and idiomatic as it is written. I see no reason to change that just because the type checker can't handle it. At this point, mypy has arguably done more harm than good to this project, by frequently breaking CI for no reason and exposing O(1) bugs at best. So I'm very hesitant to change the API just to make it happy.

Reply:
I would say that "typing overload" is exactly what generics is, so this solution would be equivalent to using generics. Just different syntax :slightly_smiling_face: 

I don't think it's worth changing either. Just commenting on that comment in the code, that there is a way to have multiple return types with `Union`, but it's not really useful because it doesn't really help you, type-check-wise or code-completion-wise.